### PR TITLE
AclStorage dependency injection

### PR DESCRIPTION
**Pull request by mlepage-google**
#### Problem
#14443

#### Change overview
Make AclStorage into an inteface and inject it into Server via initParams.

#### Testing
- Build and run (chip-all-clusters-app and chip-tool) on Linux x64.
- Verify that 2 fabrics with 2 ACL entries each survive restarting chip-all-clusters-app.

### CODE REVIEW

File: src/app/server/AclStorage.h [hunk 1]
```diff
@@ -138,12 +138,19 @@ class AclStorage
...
+    virtual ~AclStorage() = default;
+
     /**
-     * Initialize must be called. It loads ACL entries for all fabrics from persistent storage,
-     * then installs a listener for the access control system module to maintain ACL entries in
-     * persistent storage so they remain in sync with entries in the access control system module.
+     * Initialize should be called after chip::Access::AccessControl is initialized.
+     *
+     * Implementations should take this opportunity to populate AccessControl with ACL entries
+     * loaded from persistent storage. A half-open range of fabrics (first to last) is provided
```
Comment (bzbarsky-apple) [inline src/app/server/AclStorage.h:147]:
```suggestion
     * loaded from persistent storage. A half-open range of fabrics [first, last) is provided
```

File: src/app/server/ExampleAclStorage.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#include <app/server/AclStorage.h>
+
+namespace chip {
+namespace app {
+
+class ExampleAclStorage : public AclStorage
+{
+public:
+    /**
+     * Initialize must be called. It loads ACL entries for all fabrics (half-open range first to
```
Comment (bzbarsky-apple) [inline src/app/server/ExampleAclStorage.h:29]:
```suggestion
     * Initialize must be called. It loads ACL entries for all fabrics in the half-open range [first, last)
```

File: src/app/server/ExampleAclStorage.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+     * last) from persistent storage, then installs a listener for the access control system module
```
Comment (bzbarsky-apple) [inline src/app/server/ExampleAclStorage.h:30]:
```suggestion
     * from persistent storage, then installs a listener for the access control system module
```

File: src/app/server/ExampleAclStorage.cpp [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+#include <app/server/ExampleAclStorage.h>
+
+#include <lib/support/DefaultStorageKeyAllocator.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::Access;
+
+using EncodableEntry   = AclStorage::EncodableEntry;
```
Comment (bzbarsky-apple) [inline src/app/server/ExampleAclStorage.cpp:26]: But isn't  using `AclStorage::EncodableEntry` something specific to this storage implementation?  That is, why is that type on AclStorage and not on this subclass?

Comment (mlepage-google) [reply src/app/server/ExampleAclStorage.cpp:26]: Those are helper classes provided here. They don't have to be used by any derived implementation. But they are used by the cluster code (which isn't involved with persistent storage, but does encode/decode ACLs). That code could be broken out into another place, but I'm in favour of simplicity.

File: src/app/server/ExampleAclStorage.h (hunk 2 above)
Comment (mlepage-google) [reply src/app/server/ExampleAclStorage.h:29]: Do the square brackets work with whatever Javadoc-like thing we're using, they won't confuse it? I thought about using them but thought it better to just write it in English (half-open range first to last, people can search what half-open means).

Comment (bzbarsky-apple) [reply src/app/server/ExampleAclStorage.h:29]: The problem I was having is that "half-open" doesn't say which end is closed and which is open....

Comment (mlepage-google) [reply src/app/server/ExampleAclStorage.h:29]: Ah, I guess that's true, though it's conventional in C++ land that it's the way it's always about iterators. Honestly I should be able to put no docs there and it should be clear how it works. :-)

Comment (mlepage-google) [reply src/app/server/ExampleAclStorage.h]: I clarified how the half-open range works in the base function, then avoided repeating that in the derived function, and instead just explained what it actually does in the implementation.

File: src/app/server/AclStorage.h (hunk 1 above)
Comment (mlepage-google) [reply src/app/server/AclStorage.h]: Clarified in latest revision.

File: src/app/server/ExampleAclStorage.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+#pragma once
+
+#include <app/server/AclStorage.h>
+
+namespace chip {
+namespace app {
+
+class ExampleAclStorage : public AclStorage
```
Comment (tcarmelveilleux) [inline src/app/server/ExampleAclStorage.h:25]: Suggest rename from `ExampleAclStorage` to `DefaultAclStorage`

**Review by tcarmelveilleux — APPROVED** [review]

File: src/app/server/ExampleAclStorage.h (hunk 5 above)
Comment (mlepage-google) [reply src/app/server/ExampleAclStorage.h]: Sure.

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**